player/lua/stats.lua

```diff
@@ -0,0 +1,734 @@
+        end
+    end
+end
+
+
+-- Call the function for `page` and print it to OSD
+local function print_page(page)
+    if o.persistent_overlay then
+        mp.set_osd_ass(0, 0, pages[page].f())
+    else
+        mp.osd_message(pages[page].f(), display_timer.oneshot and o.duration or o.redraw_delay + 1)
+    end
+end
+
+
+local function clear_screen()
+    if o.persistent_overlay then mp.set_osd_ass(0, 0, "") else mp.osd_message("", 0) end
+end
+
+
+-- Add keybindings for every page
+local function add_page_bindings()
+    local function a(k)
+        return function()
+            curr_page = k
+            print_page(k)
+            if display_timer.oneshot then display_timer:kill() ; display_timer:resume() end
+        end
+    end
+    for k, _ in pairs(pages) do
+        mp.add_forced_key_binding(k, k, a(k), {repeatable=true})
+    end
+end
+
+
+-- Remove keybindings for every page
+local function remove_page_bindings()
+    for k, _ in pairs(pages) do
+        mp.remove_key_binding(k)
+    end
+end
+
+
+local function process_key_binding(oneshot)
+    -- Stats are already being displayed
+    if display_timer:is_enabled() then
+        -- Previous and current keys were oneshot -> restart timer
+        if display_timer.oneshot and oneshot then
+            display_timer:kill()
+            print_page(curr_page)
+            display_timer:resume()
+        -- Previous and current keys were toggling -> end toggling
+        elseif not display_timer.oneshot and not oneshot then
+            display_timer:kill()
+            clear_screen()
+            remove_page_bindings()
+            if recorder then
+                mp.unregister_event(recorder)
+                recorder = nil
+            end
+        end
+    -- No stats are being displayed yet
+    else
+        if not oneshot and (o.plot_vsync_jitter or o.plot_vsync_ratio) then
+            recorder = record_data(o.skip_frames)
+            mp.register_event("tick", recorder)
+        end
+        display_timer:kill()
+        display_timer.oneshot = oneshot
+        display_timer.timeout = oneshot and o.duration or o.redraw_delay
+        add_page_bindings()
+        print_page(curr_page)
+        display_timer:resume()
+    end
+end
+
+
+-- Create the timer used for redrawing (toggling) or clearing the screen (oneshot)
+-- The duration here is not important and always set in process_key_binding()
+display_timer = mp.add_periodic_timer(o.duration,
+    function()
+        if display_timer.oneshot then
+            display_timer:kill() ; clear_screen() ; remove_page_bindings()
+        else
+            print_page(curr_page)
+        end
+    end)
+display_timer:kill()
+
+-- Single invocation key binding
+mp.add_key_binding(o.key_oneshot, "display-stats", function() process_key_binding(true) end,
+    {repeatable=true})
+
+-- Toggling key binding
+mp.add_key_binding(o.key_toggle, "display-stats-toggle", function() process_key_binding(false) end,
+    {repeatable=false})
```

Comment:
Not that it matters much IMO, but why do you disable repeat here?

Reply:
There's no point. There's no advantage to gain from key repeat here but there are plenty of counterintuitive disadvantages.